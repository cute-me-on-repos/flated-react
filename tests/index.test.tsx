import { render } from "@testing-library/react";
import FlatedReact from "../src"; // Adjust the import based on your project structure

// Mock components for testing
const MockComponentA: React.FC<
  React.PropsWithChildren<{ message: string }>
> = ({ message, children }) => (
  <div>
    A: {message}
    {children}
  </div>
);

const MockComponentB: React.FC<React.PropsWithChildren<{ count: number }>> = ({
  count,
  children,
}) => (
  <div>
    B: {count}
    {children}
  </div>
);

let componentCIndex = 0;
const MockComponentC: React.FC<React.PropsWithChildren> = ({ children }) => (
  <div>
    C: {++componentCIndex}
    {children}
  </div>
);

describe("FlatedReact", () => {
  it("renders nested components correctly", () => {
    const { queryAllByLabelText, container } = render(
      <FlatedReact.Wrap
        components={[
          FlatedReact.Load(MockComponentA, { message: "Hello" }),
          FlatedReact.Load(MockComponentB, { count: 42 }),
          MockComponentC,
        ]}
      >
        <span>Children Content</span>
      </FlatedReact.Wrap>
    );

    expect(queryAllByLabelText("A: Hello")).toBeTruthy();
    expect(queryAllByLabelText("B: 42")).toBeTruthy();
    expect(queryAllByLabelText("C")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("renders components with default props", () => {
    const { queryAllByLabelText, container } = render(
      <FlatedReact.Wrap
        components={[
          FlatedReact.Load(MockComponentA, { message: "Default Message" }),
          FlatedReact.Load(MockComponentC),
        ]}
      >
        <span>Default Children</span>
      </FlatedReact.Wrap>
    );

    expect(queryAllByLabelText("A: Default Message")).toBeTruthy();
    expect(queryAllByLabelText("C")).toBeTruthy();
    expect(queryAllByLabelText("Default Children")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("renders without props and `FlatedReact.Load` correctly", () => {
    const { queryAllByLabelText, container } = render(
      <FlatedReact.Wrap
        components={[
          FlatedReact.Load(MockComponentC),
          [MockComponentC],
          [MockComponentC, undefined],
          [MockComponentC, null],
          [MockComponentC, {}],
        ]}
      >
        <span>No Props Children</span>
      </FlatedReact.Wrap>
    );

    expect(queryAllByLabelText("C")).toBeTruthy();
    expect(queryAllByLabelText("No Props Children")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
