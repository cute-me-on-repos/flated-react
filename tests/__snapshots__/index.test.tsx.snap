// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatedReact renders components with default props 1`] = `
<div>
  <div>
    A: 
    Default Message
    <div>
      C: 
      2
      <span>
        Default Children
      </span>
    </div>
  </div>
</div>
`;

exports[`FlatedReact renders nested components correctly 1`] = `
<div>
  <div>
    A: 
    Hello
    <div>
      B: 
      42
      <div>
        C: 
        1
        <span>
          Children Content
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`FlatedReact renders without props and \`FlatedReact.Load\` correctly 1`] = `
<div>
  <div>
    C: 
    3
    <div>
      C: 
      4
      <div>
        C: 
        5
        <div>
          C: 
          6
          <div>
            C: 
            7
            <span>
              No Props Children
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
